README.md
把蚂蚁金服开放平台的sdk放到了composer里面,

有问题请邮件联系我： 76762@qq.com ~

# 百度Tongji Api SDK  composer版

[![Latest Version on Packagist](https://img.shields.io/packagist/v/adzon/openalipaysdk.svg?style=flat-square)](https://packagist.org/packages/adzon/openalipaysdk)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/adzon/openalipaysdk/master.svg?style=flat-square)](https://travis-ci.org/adzon/openalipaysdk)
[![StyleCI](https://styleci.io/repos/:style_ci_id/shield)](https://styleci.io/repos/:style_ci_id)
[![Quality Score](https://img.shields.io/scrutinizer/g/adzon/openalipaysdk.svg?style=flat-square)](https://scrutinizer-ci.com/g/adzon/openalipaysdk)
[![Code Coverage](https://img.shields.io/scrutinizer/coverage/g/adzon/openalipaysdk/master.svg?style=flat-square)](https://scrutinizer-ci.com/g/adzon/openalipaysdk/?branch=master)
[![Total Downloads](https://img.shields.io/packagist/dt/adzon/openalipaysdk.svg?style=flat-square)](https://packagist.org/packages/adzon/openalipaysdk)

本项目可以在 Laravel 5.4中使用 [蚂蚁金服开放平台](http://www.baidu.com/?wd=蚂蚁金服开放平台) 的 API。

## Contents

- [安装说明](#安装说明)
- [使用说明](#使用说明)
- [更新日志](#更新日志)
- [Credits](#credits)
- [License](#license)

## 安装说明

```bash
composer require adzon/openalipaysdk
```

## 使用说明

demo文件内容参见 demo/demo.php

## 更新日志

请直接参见commit提交～～～

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [adzon](https://github.com/adzon)
- [执行力就是挨个搞](http://www.huleping.com/)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes will be documented in this file

## 1.0.0 - 2017-06-08

- 构建第一个版本咯~~~

反正懒得写更新日志, 看 commit 提交好了...

src/test/TestImage.php
<?php
/**
 * Alipay.com Inc.
 * Copyright (c) 2004-2014 All Rights Reserved.
 */



namespace Alipay\Api;


header("Content-type: text/html; charset=gbk");

/**
 *
 * @author wangYuanWai
 * @version $Id: Test.hp, v 0.1 Aug 6, 2014 4:20:17 PM yikai.hu Exp $
 */
class TestImage{


	public $partner_public_key  = "MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCitD16CypwZILTpdJL8nPV9rVFHYf5UWa/URNX6469mbQLpWfjKM/VSWRXsNVGSM3itOO/KG2Pw4x5g9xjH6iaE4LlaidjBIPpifISSlnpbyi4HxQTZYgMPv/TuiWofUN5kcwg/KQAQxB2OwTOeFu2i3LhqSCDmv6koTvHW15/hQIDAQAB";
	public $alipay_public_key   = "MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDIgHnOn7LLILlKETd6BFRJ0GqgS2Y3mn1wMQmyh9zEyWlz5p1zrahRahbXAfCfSqshSNfqOmAQzSHRVjCqjsAw1jyqrXaPdKBmr90DIpIxmIyKXv4GGAkPyJ/6FTFY99uhpiq0qadD/uSzQsefWo0aTvP/65zi3eof7TcZ32oWpwIDAQAB";
		//公用变量
	public $serverUrl = 'http://publicexprod.d5336aqcn.alipay.net/chat/multimedia.do';//'http://publicexprod.d5336aqcn.alipay.net/chat/multimedia.do';//'http://i.com/works/photo-sdk/_data/1.jpg';//"http://i.com/works/photo-sdk/_data/publicexprod.php";//"http://publicexprod.d5336aqcn.alipay.net/chat/multimedia.do";
	public $appId = "2013121100055554";

	public $partner_private_key = 'MIICdwIBADANBgkqhkiG9w0BAQEFAASCAmEwggJdAgEAAoGBAKK0PXoLKnBkgtOl0kvyc9X2tUUdh/lRZr9RE1frjr2ZtAulZ+Moz9VJZFew1UZIzeK0478obY/DjHmD3GMfqJoTguVqJ2MEg+mJ8hJKWelvKLgfFBNliAw+/9O6Jah9Q3mRzCD8pABDEHY7BM54W7aLcuGpIIOa/qShO8dbXn+FAgMBAAECgYA8+nQ380taiDEIBZPFZv7G6AmT97doV3u8pDQttVjv8lUqMDm5RyhtdW4n91xXVR3ko4rfr9UwFkflmufUNp9HU9bHIVQS+HWLsPv9GypdTSNNp+nDn4JExUtAakJxZmGhCu/WjHIUzCoBCn6viernVC2L37NL1N4zrR73lSCk2QJBAPb/UOmtSx+PnA/mimqnFMMP3SX6cQmnynz9+63JlLjXD8rowRD2Z03U41Qfy+RED3yANZXCrE1V6vghYVmASYsCQQCoomZpeNxAKuUJZp+VaWi4WQeMW1KCK3aljaKLMZ57yb5Bsu+P3odyBk1AvYIPvdajAJiiikRdIDmi58dqfN0vAkEAjFX8LwjbCg+aaB5gvsA3t6ynxhBJcWb4UZQtD0zdRzhKLMuaBn05rKssjnuSaRuSgPaHe5OkOjx6yIiOuz98iQJAXIDpSMYhm5lsFiITPDScWzOLLnUR55HL/biaB1zqoODj2so7G2JoTiYiznamF9h9GuFC2TablbINq80U2NcxxQJBAMhw06Ha/U7qTjtAmr2qAuWSWvHU4ANu2h0RxYlKTpmWgO0f47jCOQhdC3T/RK7f38c7q8uPyi35eZ7S1e/PznY=';

	public $format = "json";
	public $charset = "GBK";



	function __construct(){

	}

	public function load() {
		$alipayClient = new AlipayMobilePublicMultiMediaClient(
			$this -> serverUrl,
			$this -> appId,
			$this -> partner_private_key,
			$this -> format,
			$this -> charset
		);
		$response = null;
		$outputStream = null;
		$request = $alipayClient -> getContents() ;

		//200
		//echo( '状态码：'. $request -> getCode() .', ');
		//echo '<hr /><br /><br /><br />';

		$fileType = $request -> getType();
		//echo( '类型：'. $fileType .', ');
		if( $fileType == 'text/plain'){
			//出错，返回 json
			echo $request -> getBody();

		}else{

			$type = $request -> getFileSuffix( $fileType );

			//echo $this -> getParams();
			//exit();

			//返回 文件流
			header("Content-type: ". $fileType ); //类型


			header("Accept-Ranges: bytes");//告诉客户端浏览器返回的文件大小是按照字节进行计算的
			header("Accept-Length: ". $request -> getContentLength() );//文件大小
			header("Content-Length: ". $request -> getContentLength() );//文件大小
			header('Content-Disposition: attachment; filename="'. time() .'.'. $type .'"'); //文件名
			echo $request -> getBody() ;
			exit ( ) ;
		}

		//echo( '内容： , '. $request -> getContentLength()  );

		//echo '<hr /><br /><br /><br />';
		//echo  '参数：<pre>';

		//echo ($request -> getParams());

		//echo '</pre>' ;
	}
}





//  测试
$test1 = new TestImage();
$test1 -> load();


src/AopEncrypt.php
<?php
/**
 *   加密工具类
 *
 * User: jiehua
 * Date: 16/3/30
 * Time: 下午3:25
 */
namespace Alipay\Api;
/**
 * 加密方法
 * @param string $str
 * @return string
 */
 function encrypt($str,$screct_key){
	//AES, 128 模式加密数据 CBC
	$screct_key = base64_decode($screct_key);
	$str = trim($str);
	$str = addPKCS7Padding($str);
	$iv = mcrypt_create_iv(mcrypt_get_iv_size(MCRYPT_RIJNDAEL_128,MCRYPT_MODE_CBC),1);
	$encrypt_str =  mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $screct_key, $str, MCRYPT_MODE_CBC);
	return base64_encode($encrypt_str);
}

/**
 * 解密方法
 * @param string $str
 * @return string
 */
 function decrypt($str,$screct_key){
	//AES, 128 模式加密数据 CBC
	$str = base64_decode($str);
	$screct_key = base64_decode($screct_key);
	$iv = mcrypt_create_iv(mcrypt_get_iv_size(MCRYPT_RIJNDAEL_128,MCRYPT_MODE_CBC),1);
	$encrypt_str =  mcrypt_decrypt(MCRYPT_RIJNDAEL_128, $screct_key, $str, MCRYPT_MODE_CBC);
	$encrypt_str = trim($encrypt_str);

	$encrypt_str = stripPKSC7Padding($encrypt_str);
	return $encrypt_str;

}

/**
 * 填充算法
 * @param string $source
 * @return string
 */
function addPKCS7Padding($source){
	$source = trim($source);
	$block = mcrypt_get_block_size(MCRYPT_RIJNDAEL_128, MCRYPT_MODE_CBC);

	$pad = $block - (strlen($source) % $block);
	if ($pad <= $block) {
		$char = chr($pad);
		$source .= str_repeat($char, $pad);
	}
	return $source;
}
/**
 * 移去填充算法
 * @param string $source
 * @return string
 */
function stripPKSC7Padding($source){
	$source = trim($source);
	$char = substr($source, -1);
	$num = ord($char);
	if($num==62)return $source;
	$source = substr($source,0,-$num);
	return $source;
}

src/AlipayMobilePublicMultiMediaExecute.php
<?php

/**
 * 多媒体文件客户端
 * @author yuanwai.wang
 * @version $Id: AlipayMobilePublicMultiMediaExecute.php, v 0.1 Aug 15, 2014 10:19:01 AM yuanwai.wang Exp $
 */

//namespace alipay\api ;
namespace Alipay\Api;


class AlipayMobilePublicMultiMediaExecute{

	private $code = 200 ;
	private $msg = '';
	private $body = '';
	private $params = '';

	private $fileSuffix = array(
		"image/jpeg" 		=> 'jpg', //+
		"text/plain"		=> 'text'
	);

	/*
	 * @$header : 头部
	 * */
	function __construct( $header, $body, $httpCode ){
		$this -> code = $httpCode;
		$this -> msg = '';
		$this -> params = $header ;
		$this -> body = $body;
	}

	/**
	 *
	 * @return text | bin
	 */
	public function getCode(){
		return $this -> code ;
	}

	/**
	 *
	 * @return text | bin
	 */
	public function getMsg(){
		return $this -> msg ;
	}

	/**
	 *
	 * @return text | bin
	 */
	public function getType(){
		$subject = $this -> params ;
		$pattern = '/Content\-Type:([^;]+)/';
		preg_match($pattern, $subject, $matches);
		if( $matches ){
			$type = $matches[1];
		}else{
			$type = 'application/download';
		}

		return str_replace( ' ', '', $type );
	}

	/**
	 *
	 * @return text | bin
	 */
	public function getContentLength(){
		$subject = $this -> params ;
		$pattern = '/Content-Length:\s*([^\n]+)/';
		preg_match($pattern, $subject, $matches);
		return (int)( isset($matches[1] ) ? $matches[1]  : '' );
	}


	public function getFileSuffix( $fileType ){
		$type = isset( $this -> fileSuffix[ $fileType ] ) ? $this -> fileSuffix[ $fileType ] : 'text/plain' ;
		if( !$type ){
			$type = 'json';
		}
		return $type;
	}



	/**
	 *
	 * @return text | bin
	 */
	public function getBody(){
		//header('Content-type: image/jpeg');
		return $this -> body ;
	}

	/**
	 * 获取参数
	 * @return text | bin
	 */
	public function getParams(){
		return $this -> params ;
	}


}


src/request/AlipayAccountExrateAdviceAcceptRequest.php
<?php
/**
 * ALIPAY API: alipay.account.exrate.advice.accept request
 *
 * @author auto create
 * @since 1.0, 2016-05-23 14:55:42
 */
namespace Alipay\Api;


class AlipayAccountExrateAdviceAcceptRequest
{
	/** 
	 * 标准的兑换交易受理接口
	 **/
	private $bizContent;

	private $apiParas = array();
	private $terminalType;
	private $terminalInfo;
	private $prodCode;
	private $apiVersion="1.0";
	private $notifyUrl;
	private $returnUrl;
    private $needEncrypt=false;

	
	public function setBizContent($bizContent)
	{
		$this->bizContent = $bizContent;
		$this->apiParas["biz_content"] = $bizContent;
	}

	public function getBizContent()
	{
		return $this->bizContent;
	}

	public function getApiMethodName()
	{
		return "alipay.account.exrate.advice.accept";
	}

	public function setNotifyUrl($notifyUrl)
	{
		$this->notifyUrl=$notifyUrl;
	}

	public function getNotifyUrl()
	{
		return $this->notifyUrl;
	}

	public function setReturnUrl($returnUrl)
	{
		$this->returnUrl=$returnUrl;
	}

	public function getReturnUrl()
	{
		return $this->returnUrl;
	}

	public function getApiParas()
	{
		return $this->apiParas;
	}

	public function getTerminalType()
	{
		return $this->terminalType;
	}

	public function setTerminalType($terminalType)
	{
		$this->terminalType = $terminalType;
	}

	public function getTerminalInfo()
	{
		return $this->terminalInfo;
	}

	public function setTerminalInfo($terminalInfo)
	{
		$this->terminalInfo = $terminalInfo;
	}

	public function getProdCode()
	{
		return $this->prodCode;
	}

	public function setProdCode($prodCode)
	{
		$this->prodCode = $prodCode;
	}

	public function setApiVersion($apiVersion)
	{
		$this->apiVersion=$apiVersion;
	}

	public function getApiVersion()
	{
		return $this->apiVersion;
	}

  public function setNeedEncrypt($needEncrypt)
  {

     $this->needEncrypt=$needEncrypt;

  }

  public function getNeedEncrypt()
  {
    return $this->needEncrypt;
  }

}


src/AlipayMobilePublicMultiMediaClient.php
<?php

/**
 * 多媒体文件客户端
 * @author yikai.hu
 * @version $Id: AlipayMobilePublicMultiMediaClient.php, v 0.1 Aug 15, 2014 10:19:01 AM yikai.hu Exp $
 */

//namespace alipay\api ;

//include("AlipayMobilePublicMultiMediaExecute.php");

namespace Alipay\Api;


class AlipayMobilePublicMultiMediaClient{

    private				$DEFAULT_CHARSET = 'UTF-8';
    private				$METHOD_POST     = "POST";
    private				$METHOD_GET      = "GET";
	private				$SIGN			= 'sign'; //get name

	private				$timeout = 10 ;// 超时时间
    private				$serverUrl;
    private				$appId;
    private				$privateKey;
    private				$prodCode;
    private				$format          = 'json'; //todo
    private				$sign_type       = 'RSA'; //todo

	private				$charset;
	private				$apiVersion    = "1.0";
	private				$apiMethodName = "alipay.mobile.public.multimedia.download";
	private				$media_id = "L21pZnMvVDFQV3hYWGJKWFhYYUNucHJYP3Q9YW13ZiZ4c2lnPTU0MzRhYjg1ZTZjNWJmZTMxZGJiNjIzNDdjMzFkNzkw575";
	//此处写死的，实际开发中，请传入

    private				$connectTimeout  = 3000;
    private				$readTimeout     = 15000;



	function __construct($serverUrl = '', $appId = '', $partner_private_key = '', $format = '', $charset = 'GBK'){
		$this -> serverUrl = $serverUrl;
		$this -> appId = $appId;
		$this -> privateKey = $partner_private_key;
		$this -> format = $format;
		$this -> charset = $charset;
	}

	/**
	 * getContents 获取网址内容
	 * @param $request
	 * @return text | bin
	 */
	public function getContents(){
		//自己的服务器如果没有 curl，可用：fsockopen() 等


		//1:
		//2： 私钥格式
		$datas = array(
			"app_id" 		=> $this -> appId,
			"method" 		=> $this -> METHOD_POST,
			"sign_type" 	=> $this -> sign_type,
			"version" 		=> $this -> apiVersion,
			"timestamp" 	=> date('Y-m-d H:i:s')  ,//yyyy-MM-dd HH:mm:ss
			"biz_content" 	=> '{"mediaId":"'. $this -> media_id  .'"}',
			"charset" 		=> $this -> charset
		);



		//要提交的数据
		$data_sign = $this -> buildGetUrl( $datas );

		$post_data = $data_sign;
		//初始化 curl
		$ch = curl_init();
		//设置目标服务器
		curl_setopt($ch, CURLOPT_URL, $this -> serverUrl );
		curl_setopt($ch, CURLOPT_HEADER, TRUE);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
		//超时时间
		curl_setopt($ch, CURLOPT_TIMEOUT, $this-> timeout);

		if( $this-> METHOD_POST == 'POST'){
			// post数据
			curl_setopt($ch, CURLOPT_POST, 1);
			// post的变量
			curl_setopt($ch, CURLOPT_POSTFIELDS, $post_data);
		}




		$output = curl_exec($ch);
		$httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
		curl_close($ch);

		echo $output;

		//分离头部
		//list($header, $body) = explode("\r\n\r\n", $output, 2);
		$datas = explode("\r\n\r\n", $output, 2);
		$header = $datas[0];

		if( $httpCode == '200'){
			$body = $datas[1];
		}else{
			$body = '';

		}




		return $this -> execute( $header, $body, $httpCode );
	}

	/**
	 *
	 * @param $request
	 * @return text | bin
	 */
	public function execute( $header = '', $body = '', $httpCode = '' ){
		$exe = new AlipayMobilePublicMultiMediaExecute( $header, $body, $httpCode );
		return $exe;
	}

	public function buildGetUrl( $query = array() ){

		if( ! is_array( $query ) ){
			//exit;
		}

		//排序参数，
		$data = $this -> buildQuery( $query );



		// 私钥密码
		$passphrase = '';
		$key_width = 64;

		//私钥
		$privateKey = $this -> privateKey;
		$p_key = array();
		//如果私钥是 1行
		if( ! stripos( $privateKey, "\n" )  ){
			$i = 0;
			while( $key_str = substr( $privateKey , $i * $key_width , $key_width) ){
				$p_key[] = $key_str;
				$i ++ ;
			}
		}else{
			//echo '一行？';
		}
		$privateKey = "-----BEGIN RSA PRIVATE KEY-----\n" . implode("\n", $p_key) ;
		$privateKey = $privateKey ."\n-----END RSA PRIVATE KEY-----";

//		echo "\n\n私钥:\n";
//		echo( $privateKey );
//		echo "\n\n\n";

		//私钥
		$private_id = openssl_pkey_get_private( $privateKey , $passphrase);


		// 签名
		$signature = '';

		if("RSA2"==$this->sign_type){

			openssl_sign($data, $signature, $private_id, OPENSSL_ALGO_SHA256 );
		}else{

			openssl_sign($data, $signature, $private_id, OPENSSL_ALGO_SHA1 );
		}

		openssl_free_key( $private_id );

		//加密后的内容通常含有特殊字符，需要编码转换下
		$signature = base64_encode($signature);

		$signature = urlencode( $signature );

		//$signature = 'XjUN6YM1Mc9HXebKMv7GTLy7gmyhktyOgKk2/Jf+cz4DtP6udkzTdpkjW2j/Z4ZSD7xD6CNYI1Spz4yS93HPT0a5X9LgFWYY8SaADqe+ArXg+FBSiTwUz49SE//Xd9+LEiIRsSFkbpkuiGoO6mqJmB7vXjlD5lx6qCM3nb41wb8=';

		$out = $data .'&'. $this -> SIGN .'='. $signature;

//		echo "\n\n 加密后:\n";
//		echo( $out );
//		echo "\n\n\n";

		return $out ;
	}

	/*
	 * 查询参数排序 a-z
	 * */
	public function buildQuery( $query ){
		if ( !$query ) {
			return null;
		}

//将要 参数 排序
		ksort( $query );

		//重新组装参数
		$params = array();
		foreach($query as $key => $value){
			$params[] = $key .'='. $value ;
		}
		$data = implode('&', $params);

		return $data;

	}














}
